feat(common): Add an`Intl` implementation for the i18n sub-system.

JeanMeche opened the pull request:
Angular's i18n subsystem is responsible for providing localised formatting of numbers, percents, percentages and dates. 
This feature replaces the default implementation of the i18N subsystem. 

The commits add a new implementation that relies on the platform-provided [`Intl` API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl). 

With this new implementation relying on a platform API, it is not necessary anymore the load locale files (See #20487) . 

The new implementation is available via an opt-in : `useIntlImplementation()`. 

This feature fixes the following issues: 
* angular/angular#20487
* angular/angular#33803 
* angular/angular#39606
* angular/angular#42315
* #48279
* angular/angular#48697
* angular/angular#49143
* #49540
* angular/angular#52177
* angular/angular#54114
* angular/angular#54320
* #56907

Closes #54470

alan-agius4 commented on packages/core/src/i18n/locale_data_api.ts:
```diff
@@ -15,6 +15,8 @@ import localeEn from './locale_en';
  */
 let LOCALE_DATA: {[localeId: string]: any} = {};
 
+export const PLURALS_DATA: {[localeId: string]: Intl.PluralRules} = {};
```
```suggestion
const PLURALS_DATA: {[localeId: string]: Intl.PluralRules} = {};
```

JeanMeche committed:
feat(common): add `Intl` implementation to format numbers, percents and currencies

This commit adds an `Intl` implementation to format numbers, percents and currencies.
The implementation is based on the `Intl` API and allows to drop the static imports for the locale date files.

fixes #33803

JeanMeche committed:
feat(common): use the `Intl` API to for plurals

With this commit, plurals formatting does not depend on the CLDR locale data anymore.

JeanMeche committed:
feat(common): add an `Intl` API implementation for date formatting.

With this commit, the Intl implementation because the default one and doesn't require the CLDR locale data imports anymore.

Opt-out is possible by invoking `useLegacyDateFormatting()` ahead of bootstrap.

BREAKING CHANGE: Some custom date format aren't supported any more (`cccccc`, `EEEEEE`, `aaaaa`, `b` to `bbbbb` and `B` to `BBBBB`)
The `DatePipe` will not support offset timezone anymore, use IANA timezones instead .

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/OCL:626201892:BASE:626239048:1713500111099:42ea321c).

AndrewKushnir commented:
FYI, TGP contains a number of failures, which indicates that we should re-consider the rollout strategy (I've shared more info offline).

devversion commented on packages/common/src/i18n/intl_helpers.ts:
```diff
@@ -0,0 +1,150 @@
+/** Numbers */
+
+// For values >= 1e21 the Intl implementation will return
+// a string as scientific notation like the native api does
```
Can you elaborate more in this comment? This is for backwards compatibility right?

JeanMeche replied:
Let me know if the change makes it more clear.

devversion commented on packages/common/src/i18n/intl_helpers.ts:
```diff
@@ -0,0 +1,150 @@
+      // TODO: create a runtime error
+      throw new Error(`${digitsInfo} is not a valid digit info`);
+    }
+    const [, minIntPart, , minFractionPart, , maxFractionPart] = parts; // yes we want indices 1,3,5
```
Probably better to use individual accesses. The comment is not super useful

devversion commented on packages/common/src/i18n/intl_helpers.ts:
```diff
@@ -0,0 +1,150 @@
+    minimumFractionDigits: undefined | number,
+    maximumFractionDigits: undefined | number;
+  if (digitsInfo) {
+    const parts = digitsInfo.match(NUMBER_FORMAT_REGEXP);
```
Is there a way to share this logic with the legacy logic? that would also ease review I think

devversion commented on packages/common/src/i18n/intl_helpers.ts:
```diff
@@ -0,0 +1,150 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+const NUMBER_FORMAT_REGEXP = /^(\d+)?\.((\d+)(-(\d+))?)?$/;
```
Maybe `DIGITS_INFO_REGEX` and link to something like https://angular.io/api/common/DecimalPipe#digitsinfo

devversion commented on packages/common/src/i18n/intl_helpers.ts:
```diff
@@ -0,0 +1,150 @@
+    parseDigitInfo(digitsInfo);
+
+  // convert Pipe display option to Intl display option
+  if (displayOrCurrency === 'symbol-narrow') {
```
Would the pipe be able to directly handle this?

devversion commented on packages/common/src/i18n/intl_helpers.ts:
```diff
@@ -0,0 +1,150 @@
+  // TODO: remove `as any` when ts target is updated to at least ES2022
+  // Support for supportedValuesOf is M99/Fx93/Safari 15.4
+  const isCurrencySupported =
+    (Intl as any).supportedValuesOf?.(CURRENCY).includes(currencyCode) ?? true;
```
Maybe rather augment `Intl` for some type safety?

JeanMeche replied:
Wdyt of adding `es2022.Intl` to our `tsconfig` ?

devversion replied:
Sounds great. It can be breaking if these types end up appearing in `d.ts` of `node_modules/@angular/core`— but if we use them internally that should be fine & desirable.

devversion commented on packages/common/src/i18n/intl_helpers.ts:
```diff
@@ -0,0 +1,150 @@
+  const formatter = Intl.NumberFormat(locale, {
+    maximumFractionDigits: isCurrencySupported
+      ? maximumFractionDigits
+      : // For unsupported currencies, we always show a minimm of 2 fraction digits
```
Nit: typo

devversion replied:
Also, why is the minimum fraction digits guarded by `isCurrencySupported`? Is this for some backwards compatibility?

devversion commented on packages/common/src/i18n/intl_helpers.ts:
```diff
@@ -0,0 +1,150 @@
+  currencyCode?: string,
+  digitsInfo?: string,
+): string {
+  const forceCurrencyDisplay = !CURRENCY_PIPE_DISPLAYS.includes(displayOrCurrency);
```
Nit: add comment on why this is done

JeanMeche replied:
Was able to drop that constant.

devversion commented on packages/common/src/i18n/intl_helpers.ts:
```diff
@@ -0,0 +1,150 @@
+    minimumFractionDigits,
+    style: CURRENCY,
+    currencyDisplay: isValidDisplay ? displayOrCurrency : undefined,
+    currency: isCurrencySupported ? currencyCode : 'USD', // USD is a placeholder, it will be replaced
```
... later by the fallback formatting?

devversion commented on packages/core/src/i18n/locale_data_api.ts:
```diff
@@ -90,18 +92,21 @@ export function getLocaleCurrencyCode(locale: string): string|null {
  */
-export function getLocalePluralCase(locale: string): (value: number) => number {
-  const data = findLocaleData(locale);
-  return data[LocaleDataIndex.PluralCase];
```
Is this a breaking change? i.e. there isn't even a way to get back to legacy CLDR here AFAICT

JeanMeche replied:
TGP returned green for this. 

Here is the support form `Intl.PluralRules.supportedLocalesOf`, it is widely within the scope of our supported browser versions. 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules/supportedLocalesOf

Depending on how strict we are on our breaking change policy, I can be considered breaking or non-breaking. Happy what you (and others) think about this.

devversion replied:
Does Google3 actually use the CLDR plural data? I don't know off hand.

JeanMeche replied:
I changed my mind on this. Let's stay on the safe side and make plural also opt-in.

devversion reviewed (commented):
Review of currency formatting + plural

JeanMeche commented on goldens/public-api/common/index.md:
```diff
@@ -167,7 +167,7 @@ export class DecimalPipe implements PipeTransform {
 
 // @public
-export function formatCurrency(value: number, locale: string, currency: string, currencyCode?: string, digitsInfo?: string): string;
+export function formatCurrency(value: number, locale: string, display: 'code' | 'symbol' | 'symbol-narrow' | string, currencyCode?: string, digitsInfo?: string): string;
```
The type is just make a bit more explicit (by including 3 specific strings)  but still includes `string`. Are you sure this is a breaking change ?

devversion replied:
Ahh, I didn't realize it still includes `string`. SGTM then.

JeanMeche commented on packages/common/test/pipes/number_pipe_spec.ts:
```diff
@@ -16,18 +22,32 @@ import localeFr from '@angular/common/locales/fr';
-
-  afterAll(() => ɵunregisterLocaleData());
+// Following ignore is to ease the review of the diff
+// prettier-ignore
```
I'll fix the formatting once the PR is merge.

JeanMeche commented on packages/common/test/i18n/format_number_spec.ts:
```diff
@@ -143,6 +176,37 @@ describe('Format number', () => {
+          // there is a non-breakable space between the currency and the value in en-US
+
+          // We know this test fails on Saucelabs/Android
+          // because it runs a rather old version of Chrome (m100)
```
Looking to make the change in #55710 but the migration to appium 2.0 is not as stragiht forward as I want it. 
I've asked some help to Joey.

JeanMeche replied:
So I finally got Saucelab running a more recent Android/Chrome version (v14, M117)  but the problem/differene still exist. 

Fwiw, I could not reproduce the issue on an Android Simulator with Chrome 120.

csvn commented:
I hope I'm not being a nuisance, but what is the status for this PR? I see it was removed from v20 milestone recently, and the last thing mentioned was in [#57916](https://github.com/angular/angular/issues/57916#issuecomment-2371581107):

> I hope that we can ship it with v19, mid November.

- Since using `Intl` is opt-in, can't this land in a minor version?
- I looked through tests and code a bit, is there any work left for this to be "done"?

I've been considering replacing `@angular/common` in our application with pipes/helpers based on `Intl` so we can format based on the users current locale instead, which is a lot messier when `registerLocaleData()` is required to use. I thought we'd wait perhaps for this PR to land though, to avoid building something custom if it will be supported in Angular. 😅

JeanMeche committed:
fixup! feat(common): add `Intl` implementation to format numbers, percents and currencies

JeanMeche committed:
fixup! feat(common): use the `Intl` API to for plurals

JeanMeche committed:
fixup! feat(common): add an `Intl` API implementation for date formatting.

mauriziocescon commented:
I wonder: any hope for v21 or no way?
